### ISSUE TITLE
Control flow on back edge of loop

### ISSUE THREAD

**Issue by ivogabe** [microsoft/TypeScript#9675]
**TypeScript Version:**  2.0 beta

**Code**

``` ts
interface A { a; x: B; }
interface B { b; x: C; }
interface C { c; x: A; }
let x: A | B | C | Function = undefined as any as A;
while (1) {
    x = x.x;
    x; // 1
}
x; // 2
```

**Expected behavior:**
The type of `x` on position 1 and 2 is `A | B | C` or `A | B | C | Function` (at least that's sound)

**Actual behavior:** 
The type of `x` on position 1 is `B` and on position 2 `A | B`. It appears that all types are not propagated properly on the back edge.

A solution might be to fall back to the declared type in such a case; at least that's sound. A better solution might require some iteration to a least fixed point. Assignments are not a problem for termination in such a loop, since they filter a finite union type. There might be some problems with type guards combined with assignments. It would not terminate if there is some type that has infinite type guards for infinite subtypes. I think that the following example demonstrates that:

``` ts
interface Wrap<T> { value: T; }
function wrapContains<U, V extends U>(w: Wrap<U>, value: V): w is Wrap<V> { ... }

let x: Wrap<any> = ...;
if (!wrapContains(x, "")) return;
// x: Wrap<string>
while (1) {
  const y = x;
  x = ...;
  if (!wrapContains(x, y)) return;
  // x: Wrap<string | Wrap<string | Wrap<string | ...>>>
}
```

In the body, the type would be infinitely expanding. A solution might be to limit the amount of iterations, and if that's exceeded, then fall back to the original type.

@ahejlsberg What do you think?

**Comment by RyanCavanaugh** [comment]
I remember Anders showing me some similar example in the initial implementation of CFA and we were pretty stumped in terms of a workable algorithm that was guaranteed to have acceptable performance in all cases.

Did you encounter this in real code, or just in exploring the CFA logic? Seeing some real-world code that would hit this would be informative.

**Comment by RyanCavanaugh** [comment]
Fun aside: is the exhaustively-correct algorithm (which would find all fixed points) Turing complete? With function overloads and string literal types I think you could represent the state in the first parameter and a finite-length tape in the remaining parameters.

**Comment by ivogabe** [comment]
I found this while browsing through the source code, and though: 'how does this handle such a case'. I tested it, and it appeared to handle these cases wrongly. A possible real world example could be the following. Imagine that the compiler would use union types for the `Node` type. A simplified example:
```typescript
type Node = Block | VariableDeclarationList | VariableDeclaration | ...;
function foo(node: VariableDeclaration) {
  let current: Node | undefined = node;
  while (current !== undefined) {
    // (*)
    current = current.parent;
  }
}
```
If I remember correctly, the type of the parent of a `VariableDeclaration` is a `VariableDeclaration List`. On the marked spot, the type of current should not be `VariableDeclaration | VariableDeclarationList`. Ideally it should be `Node`.

With only assignments, the system is not turing complete, since assignments only filter a (finite) union type, which gives only a finite amount of types/states. With type guards, there can be an infinite amount of types for a variable, for instance in the second example of the original post, so that might be turing complete. In such a case it might be desired to fall back to the original type, since that's sound.
